### ISSUE TITLE
[A11y Review] Datepicker

### ISSUE THREAD

**Issue by mendeza** [ITS-HCD/nysds#1219]
# [Datepicker](https://designsystem.ny.gov/components/datepicker/)

## Issue synopsis

1. Label + Input **interaction** not working
2. Label + Input **accessible name** not working
3. Dismiss calendar popup with <kbd>ESC</kbd> not working
4. Screen reader UX excessively verbose

---

## Jesse's audit with Claude on 2026-01-29

>Calendar popup lacks aria-controls, aria-haspopup, Escape key, focus return

---

## Native model comparison

[MDN: input type="date"](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/input/date)

```
<button>focus</button>
<label for="start">Start date:</label>
<input
  type="date"
  id="start"
  name="trip-start"
  value="2018-07-22"
  min="2018-01-01"
  max="2018-12-31" />
<button>focus</button>
```

---

### Keyboard

1. activation of associated label does not set DOM focus on component. Note: seems to be a systemic issue not unique to this component.
1. popup calendar does not close with `ESC`

### Screen reader (NVDA)

1. activation of associated label does not set SR focus on component
2. component is excessively verbose:
    >clickable  Enter in MM/DD/YYYY format  Month Schedule an appointment  spin button    Enter in MM/DD/YYYY format  0  /  Day Schedule an appointment  spin button    Enter in MM/DD/YYYY format  0  /  Year Schedule an appointment  spin button    Enter in MM/DD/YYYY format  0
    
    Compare to native model:
    >clickable spin button 07 / spin button 22 / spin button 2018

    Removing `aria-description` but leaving `aria-label` produces:
    >Month Schedule an appointment  spin button    0  /  Day Schedule an appointment  spin button    0  /  Year Schedule an appointment  spin button    0
button    Open calendar
clickable  Month Schedule an appointment  spin button    0  /  Day Schedule an appointment  spin button    0  /  Year Schedule an appointment  spin button    0
Schedule an appointment  date edit
Month Schedule an appointment  spin button  0

    This is still not acceptable verbosity.

#### Diagnosis
- Two `<label>`s per one input is an anti-pattern. Oddly, not an actual technical issue here because...
- `<label>` is not associated with input (!)
- To get around non-association, input has `aria-label` and `aria-description`. These produce an unacceptably verbose SR UX

<img alt="annotated screenshot of component code rendered in the browser inspector" width="100%" src="https://github.com/user-attachments/assets/d1497427-fe2a-4906-988c-4ce28a93b4b2" />

<details><summary><strong>Caption for annotated screenshot</strong></summary>

- A string `my-datepicker--native` is passed down from the web component to two `label` elements as the value for their `for` attributes. This string doesn't correspond to an `id` on the embedded datepicker, thus they have no effect
- There are two `label` elements targeting the same dateinput; the second is intended as the description; multiple associated labels for the same input is an anti-pattern; it would be worse if this was actually working
- In this datepicker use case, a tight association is not needed with a format description because the sub-inputs announce as "month," "day," and "year," and they are numerical spin buttons. A user literally can't enter an invalid value, thus a repeated announcement when entering each field is excessive screen reader verbosity
- In other cases where the association is needed there would be two options:
    1. The description is a `span` inside a `label`, no extra attributes
    2. The description is a `span` or `div` after the `label`, associated with `aria-describedby`
- Note: this datepicker has a hardcoded `aria-description`. This ARIA attribute is **not** preferred over `aria-describedby` for two reasons:
    1. `aria-description` is still a W3C recommendation and has less widespread support
    2.  `aria-description` is announced more assertively than `aria-describedby` and contributes to the excessive verbosity of the current implementation
- I suspect the only reason `aria-label` and `aria-description` are set on the embedded datepicker is because standard HTML associations with "for, ID" and `aria-describedby` are broken by shadowDOM

</details> 

### Voice Access

- Component is usable
   - component can be entered by invoking: a sub-string of its accessible name; "month"; "day"; "year"; "calendar"
- Component lacks dismiss by "Escape" (see above). Workaround is acceptable: "click calendar"
    - note that "calendar" is a _guess_, one of several that would be made to use all the component features, but all fairly intuitive

## Remediation goals
- Get associations to work (affects all users, including mouse)
- Dismiss with "Escape"  (affects Keyboard, Screen Reader, Voice Access)
- Simplify DOM
```
<!--EITHER:-->
<label for="ID">
    <span class="label-label">Schedule</span>
    <span class="label-desc">MM/DD/YYYY</span>
</label>
<input id="ID" type="date">

<!--OR:-->
<label for="ID">Schedule</label>
<span>MM/DD/YYYY</span>
<input id="ID" type="date">
```
- Greatly reduce SR verbosity

---

FYI @NovaCat35 @esteinborn @plasticmind

**Comment by NovaCat35** [comment]
@plasticmind The `nys-label` has issues with associating the actual native logic of the internal `<label>` with the wrapped form-related element (i.e., inputs). This is due to shadowDOM limitation, and the dev had to work around this in many instances with our current component.

I'll make a ticket out of this for the dev (me) to work on and nail out.
